### PR TITLE
Refactor the frontend.

### PR DESCRIPTION
**Pull request by dvander**
When it was time to do the 2.1 API, I figured it'd be good to simply
copy and paste the entire frontend to minimize compatibility issues.
That doesn't really work because a good deal of code (eg the database)
is shared, so really everything in one API has to be sensitive to the
overall system. Also, a great deal of delicate code (amb2.py, for
example), winds up duplicated with very little changes.

This patch splits the API-specific stuff out of generators, so we can
reduce the number of generators from 4 to 2. The API specific parts have
been renamed to what they actually are: ContextManagers. Some of that
can be shared and now exists in a base class, but most of the details of
context management will remain API specific.

An amount of the C++-specific stuff has been moved out of generators as
well, since this tended to be API specific.

While we're at it, this also simplifies the directory structure a bit.

### CODE REVIEW

File: ambuild2/context.py [hunk 1]
```diff
@@ -92,17 +93,34 @@ def reconfigure(self):
...
 
-        if api_version == '2.0':
-            from ambuild2.frontend.v2_0.amb2.gen import Generator
-        elif api_version == '2.1':
-            from ambuild2.frontend.v2_1.amb2 import Generator
-
-        gen = Generator.FromVars(self.vars, self.db, self.options.refactor)
+        from ambuild2.frontend.amb2 import Generator
+
+        if api_version >= '2.1':
```
Comment (WildCard65) [inline ambuild2/context.py:101]: I think the following is better:
```suggestion
        ContextManager = __import__('ambuild2.frontend.v{0}.context_manager'.format(api_version.replace('.', '_'))).ContextManager
```

Comment (WildCard65) [reply ambuild2/context.py:101]: Note, this suggestion should include lines 101-104

File: ambuild2/context.py [hunk 2]
```diff
@@ -92,17 +93,34 @@ def reconfigure(self):
...
+            from ambuild2.frontend.v2_1.context_manager import ContextManager
+        elif api_version >= '2.0':
+            from ambuild2.frontend.v2_0.context_manager import ContextManager
```
Comment (WildCard65) [inline ambuild2/context.py:104]: I have a better import suggestion:
```suggestion
        ContextManager = __import__('ambuild2.frontend.v{0}.context_manager'.format(api_version.replace('.', '_')),
                                    fromlist=('*',)).ContextManager
```
Idk if it's format correct though.

File: ambuild2/context.py [hunk 3]
```diff
@@ -92,17 +93,34 @@ def reconfigure(self):
...
+
+        # Backwards compatibility: for an automatic reconfigure on an older build,
+        # just assume the source path is the cwd. If the AMBuildScript suddenly
+        # has decided to depend on originalCwd, then the user may have to manually
+        # run configure.py again, until we remove configure.py entirely.
+        if 'originalCwd' in self.vars:
+            originalCwd = self.vars['originalCwd']
+        else:
+            originalCwd = self.vars['sourcePath']
```
Comment (WildCard65) [inline ambuild2/context.py:113]: 'self.vars' is just a ```dict``` no? If so, this is simpler to read is it not?
```suggestion
        originalCwd = self.vars.get('originalCwd', self.vars['sourcePath'])
```

File: ambuild2/frontend/base_generator.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+# You should have received a copy of the GNU General Public License
+# along with AMBuild. If not, see <http://www.gnu.org/licenses/>.
+
+class BaseGenerator(object):
+    def __init__(self, cm):
+        self.cm = cm
+
+    @property
+    def backend(self):
+        raise Exception('Must be implemented!')
```
Comment (WildCard65) [inline ambuild2/frontend/base_generator.py:24]: If we were using pylint for linting, you'll be chewed out for using ```Exception```
```suggestion
        raise NotImplementedError('Must be implemented!')
```
Infact, we can just turn this into an instance of ```ABCMeta``` no?
Though to do that, we require the "six" module.

File: ambuild2/frontend/base_generator.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+    def addSymlink(self, context, source, output_path):
+        raise Exception('Must be implemented!')
```
Comment (WildCard65) [inline ambuild2/frontend/base_generator.py:27]:
```suggestion
        raise NotImplementedError('Must be implemented!')
```

File: ambuild2/frontend/base_generator.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+    def addFolder(self, context, folder):
+        raise Exception('Must be implemented!')
```
Comment (WildCard65) [inline ambuild2/frontend/base_generator.py:30]:
```suggestion
        raise NotImplementedError('Must be implemented!')
```

File: ambuild2/frontend/base_generator.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+    def addCopy(self, context, source, output_path):
+        raise Exception('Must be implemented!')
```
Comment (WildCard65) [inline ambuild2/frontend/base_generator.py:33]:
```suggestion
        raise NotImplementedError('Must be implemented!')
```

File: ambuild2/frontend/base_generator.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+                        context,
+                        inputs,
+                        argv,
+                        outputs,
+                        folder = -1,
+                        dep_type = None,
+                        weak_inputs = [],
+                        shared_outputs = [],
+                        env_data = None):
+        raise Exception('Must be implemented!')
```
Comment (WildCard65) [inline ambuild2/frontend/base_generator.py:45]:
```suggestion
        raise NotImplementedError('Must be implemented!')
```

File: ambuild2/frontend/base_generator.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+    def addConfigureFile(self, context, path):
+        raise Exception('Must be implemented!')
```
Comment (WildCard65) [inline ambuild2/frontend/base_generator.py:48]:
```suggestion
        raise NotImplementedError('Must be implemented!')
```

**Comment by dvander** [comment]
Thanks for the tips, I'll keep this in mind as I rejigger stuff for 2.2.

**Pull request merged**